### PR TITLE
Signaling message reference doc

### PR DESCRIPTION
**Pull request by mcottontensor**
## Relevant components:
- [x] Signalling server
- [ ] Frontend library
- [ ] Frontend UI library
- [ ] Matchmaker
- [ ] Platform scripts
- [ ] SFU

## Problem statement:
Signaling messages are currently black magic and undocumented

## Solution
Adding a reference for every signaling message.

## Documentation
Yes

### CODE REVIEW

File: SignallingWebServer/Docs/Messages.md [hunk 1]
```diff
@@ -0,0 +1,186 @@
+## Player Sent Messages
```
Comment (lukehb) [inline SignallingWebServer/Docs/Messages.md]: - Add a top level heading called `Signalling Protocol` 
- Followed by a short explanation of the page
- Followed by a ToC the sections of this page (as it is quite a long one)

File: SignallingWebServer/Docs/Messages.md [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+### listStreamers
+
+| Param Name | Description |
+|-|-|
+
+Message is consumed by the Signaling Server. Requests the list of streamer ids currently connected to the Signaling Server. The server will reply with a [streamerList](#streamerList)
```
Comment (lukehb) [inline SignallingWebServer/Docs/Messages.md]: Signaling to `Signalling` (two l's) - for consistency with the rest of the repo.

File: SignallingWebServer/Docs/Messages.md [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+
+### subscribe
+
+| Param Name | Description |
+|---|---|
+| streamerId | The id of the stream to subscribe to. |
```
Comment (lukehb) [inline SignallingWebServer/Docs/Messages.md]: I would consider a new column in all these tables for the `type` of parameter, e.g. `string` `int` etc

File: SignallingWebServer/Docs/Messages.md [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+
+Message is consumed by the signaling server. Tells the signaling server that the player requests to subscribe to the given stream.
+
+### unsubscribe
+
+| Param Name | Description |
+|-|-|
+
+Message is consumed by the signaling server. Tells the signaling server that the player wishes to unsubscribe from the current stream. The player must have previously used the `subscribe` message for this to have any effect.
```
Comment (lukehb) [inline SignallingWebServer/Docs/Messages.md]: Where there is some dep between messages can do a section link to hyperlink to the referred section.

File: SignallingWebServer/Docs/Messages.md [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+
+### offer
+
+| Param Name | Description |
+|-|-|
+| Sdp | The WebRTC SDP package in string format |
+
+Message is forwarded to the currently subscribed to Streamer. Supplies an SDP offer to the Streamer to answer. The player must have previously used the `subscribe` message to successfully subscribe to a connected Streamer.
```
Comment (lukehb) [inline SignallingWebServer/Docs/Messages.md]: I would consider putting a glossary of terms at the start of this page:

- streamer
- player
- SFU
- signalling server
- SDP
- ICE candidate

File: SignallingWebServer/Docs/Messages.md [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+Message is consumed by the signaling server. Will print out the provided stats data on the console.
+
+## Streamer Sent Messages
+### endpointId
+
+| Param Name | Description |
+|-|-|
+| id | The id of the Streamer |
+
+Message is consumed by the Signaling Server. Specifies an id for the streamer.
```
Comment (lukehb) [inline SignallingWebServer/Docs/Messages.md]: Useful for identifying streamers in cases where they are multiple streamers (like multi-streams or running at scale in the cloud)

File: SignallingWebServer/Docs/Messages.md [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+
+### layerPreference
+
+| Param Name | Description |
+|-|-|
+| playerId | The id of the player to give the preference to |
+| spatialLayer | The index of the spatial layer to prefer |
+| temporalLayer | The index of the temporal layer to prefer |
+
+Message is forwarded to a connected SFU. Sends a preferred layer index to a connected SFU for a specified player.
```
Comment (lukehb) [inline SignallingWebServer/Docs/Messages.md]: Useful for switching between SFU quality layers to force a certain resolution/quality option either as part of UX or testing

File: SignallingWebServer/Docs/Messages.md [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+
+Message is a reply to [listStreamers](#listStreamers) from a Player. Replies with a list of currently active streamers connected to this server.
+
+### pong
+
+| Param Name | Description |
+|-|-|
+| time | The timestamp of the ping message. Will be returned in the pong message |
+
+Message is a reply to [ping](#ping) from a Streamer. Replies with the time from the ping message.
```
Comment (lukehb) [inline SignallingWebServer/Docs/Messages.md]: Seems `playerConnected` message is missing?

**Review by lukehb — CHANGES_REQUESTED** [review]
Left some comments.

Additional to those comments I think it would good to have a sequence diagram of signalling between for instance:

Player <-> SS <-> Streamer

Github supports such diagrams in markdown: https://github.blog/2022-02-14-include-diagrams-markdown-files-mermaid/

**Review by lukehb — COMMENTED** [review]
Left some comments.

Additional to those comments I think it would good to have a sequence diagram of signalling between for instance:

Player <-> SS <-> Streamer

Github supports such diagrams in markdown: https://github.blog/2022-02-14-include-diagrams-markdown-files-mermaid/

**Comment by lukehb** [comment]
@mcottontensor I would also recommend the introduction of a readme.md in the root of the SignallingWebServer directory and it can have something like the following:

# Getting Started

## Running
Run `node cirrus.js` or `platform_scripts/cmd/run_local.bat|.sh`

## Changing the frontend
Replace the contents of `/Public`

# Documentation
[Signalling Protocol](./SignallingProtocol.md)

-------------------


Once this readme.md is made it can be linked from the readme.md in the repo root.

File: SignallingWebServer/Docs/Messages.md (hunk 4 above)
Comment (mcottontensor) [reply SignallingWebServer/Docs/Messages.md]: Yeah I have been trying to do that, but in Obsidian which I'm using to write the md, I have issues with repeated headings and linking to a unique one, ie. there are multiple 'offer' headings, and I also can't seem to add invisible anchors that I can link to with unique names. Maybe the gh md interpreter can handle things differently.

File: SignallingWebServer/Docs/Messages.md (hunk 6 above)
Comment (mcottontensor) [reply SignallingWebServer/Docs/Messages.md]: Is this just suggested extra text for the description?

Comment (lukehb) [reply SignallingWebServer/Docs/Messages.md]: Yep

**Comment by mcottontensor** [comment]
> Left some comments.
> 
> Additional to those comments I think it would good to have a sequence diagram of signalling between for instance:
> 
> Player <-> SS <-> Streamer
> 
> Github supports such diagrams in markdown: https://github.blog/2022-02-14-include-diagrams-markdown-files-mermaid/

Was looking into this but a lot of messages would just be a->b or a->b->c and would then trigger different messages sent back with their own sequence.
I'm wondering if it might be worth to separate it out into its own section rather than per message... Say a section for negotiation or or stats capture, or something along those lines.

**Comment by lukehb** [comment]
I agree, on consideration a sequence diagram for negotiation would be ideal!

**Comment by lukehb** [comment]
Recent commit had code changes as opposed to pure docs, can we separate that into its own PR?

**Review by lukehb — COMMENTED** [review]
I have also think now is the time we add a section to this document called

# Version
1.0.0 - Current

Major version number - breaking protocol change such as a required new message or field or deleting an existing message.
Minor version number - independent new message.
Hotfix version - a non-breaking new field in an existing message type.

**Comment by mcottontensor** [comment]
> Recent commit had code changes as opposed to pure docs, can we separate that into its own PR?

There's no code changes? Perhaps you saw the merge from master?

**Pull request merged**